### PR TITLE
Fix ScheduleBatch req pool CPU metadata

### PR DESCRIPTION
**Pull request by luoroger37**
## Motivation

`ScheduleBatch` keeps both device-side `req_pool_indices` and a CPU mirror `req_pool_indices_cpu`. Some scheduler and bookkeeping paths, including HiSparse decode bookkeeping, consume the CPU mirror, while forward/kernel paths use the device tensor.

However, a few batch transition paths did not keep these fields consistently updated. For example, a batch could be filtered down to empty while still keeping stale request metadata, or a batch with only `req_pool_indices` populated could later enter decode/merge paths that expect `req_pool_indices_cpu` to be present. This can lead to stale request-pool metadata being observed after filtering, or crashes when merging batches with a missing CPU mirror.

This PR fixes those narrow lifecycle gaps without changing the `ScheduleBatch` data model.

## Modifications

- Populate `req_pool_indices_cpu` when constructing a HiSparse decode `ScheduleBatch` in the scheduler.
- Restore `req_pool_indices_cpu` from `req_pool_indices` before decode if the CPU mirror is missing.
- Clear request-pool and sequence-length metadata when `filter_batch()` filters all requests out of a batch.
- Restore missing CPU mirrors before `merge_batch()` concatenates request-pool indices.
- Add regression tests for:
  - restoring missing `req_pool_indices_cpu` before decode;
  - clearing stale metadata when filtering to an empty batch;
  - merging batches when either side is missing `req_pool_indices_cpu`.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #27677878103](https://github.com/sgl-project/sglang/actions/runs/27677878103)
Latest PR Test (Extra): :x: [Run #27677877286](https://github.com/sgl-project/sglang/actions/runs/27677877286)

### CODE REVIEW

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -2747,6 +2757,15 @@ def filter_batch(
             )
 
     def merge_batch(self, other: ScheduleBatch):
+        if self.req_pool_indices_cpu is None and self.req_pool_indices is not None:
```
Comment (hnyls2002) [inline python/sglang/srt/managers/schedule_batch.py:2760]: redundant guard. Please remove it

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -2690,6 +2691,15 @@ def filter_batch(
         if keep_indices is None or len(keep_indices) == 0:
             # Filter out all requests
             self.reqs = []
+            self.req_pool_indices = torch.empty(
```
Comment (hnyls2002) [inline python/sglang/srt/managers/schedule_batch.py:2694]: Also, a redundant guard.

File: python/sglang/srt/managers/schedule_batch.py
Comment (hnyls2002) [inline python/sglang/srt/managers/schedule_batch.py:1]: This file should not be changed.

File: python/sglang/srt/managers/scheduler.py [hunk 3]
```diff
@@ -2439,9 +2438,11 @@ def _build_hisparse_decode_batch(self, reqs):
             spec_algorithm=self.spec_algorithm,
         )
 
+        req_pool_indices = [r.req_pool_idx for r in reqs]
         batch.req_pool_indices = torch.tensor(
-            [r.req_pool_idx for r in reqs], dtype=torch.int64, device=device
+            req_pool_indices, dtype=torch.int64, device=device
         )
+        batch.req_pool_indices_cpu = torch.tensor(req_pool_indices, dtype=torch.int64)
```
Comment (hnyls2002) [inline python/sglang/srt/managers/scheduler.py:2445]: This is the only fix point.

**Review by hnyls2002 — APPROVED** [review]

**Comment by hnyls2002** [comment]
/rerun-test test/registered/unit/managers/test_schedule_batch_req_pool_indices.py
